File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public void fun(int open, int close, int n, String s, List<String> res){
        if(s.length() == 2*n){
            res.add(s);
            return;
        }
        if(open < n){
            fun(open+1, close, n, s+"(", res);
        }
        if(close < open){
            fun(open, close+1, n, s+")", res);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<String>();
        fun(0, 0, n, "", res);
        return res;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def fun(s, op, clo, t, n):
            if len(t) == 2*n:
                s.append(t)
                return
            if op < n:
                fun(s, op+1, clo, t+"(", n)
            if clo < op:
                fun(s, op, clo+1, t+")", n)
            
        s = []
        fun(s, 0, 0, "", n)
        return s